SSH Reconnect breaks language server

### Summary

<!-- Begin your issue with a one sentence summary -->
When a SSH connection is dropped (eg: due to a server restart) then reconnected, the language server stops working.

### Description
<!--  Describe with sufficient detail to reproduce from a clean Zed install.
  - Any code must be sufficient to reproduce (include context!)
  - Code must as text, not just as a screenshot.
  - Issues with insufficient detail may be summarily closed.
-->

- Setup a SSH target
    - Install https://github.com/bash-lsp/bash-language-server on it.
- Zed
    - Install the bash extension https://github.com/zed-extensions/bash 
    - Connect to the SSH target
    - Create a `foo.sh` file.
    - Create an unused variable (eg: `FOO=bar`): the language server will complain the variable isn't in use (good).
- Stop the SSH server (which will both kill the SSH server and the language server).
    - Zed will give a disconnect error message.
- Start the SSH server
    - And click the "Reconnect" button on Zed.
- The same `foo.sh` file will be open (good), but now there's no language server functioning, noted by the abscence of the complaint about the unused variable.

Expected Behavior: for the language server to be restarted, and continue to work.
Actual Behavior: the language server stops working completely, until Zed is restarted.

Here are the logs after the server is stopped:

```
2025-04-18T11:04:32.312314588+01:00 [WARN] ssh heartbeat: connection activity channel has been dropped. stopping.
2025-04-18T11:04:32.31240236+01:00 [ERROR] proxy process terminated unexpectedly
2025-04-18T11:04:32.312461109+01:00 [INFO] setting state to 'reconnecting'
2025-04-18T11:04:32.312523224+01:00 [INFO] Trying to reconnect to ssh server... Attempt 1
2025-04-18T11:04:32.316769212+01:00 [ERROR] Reconnect attempt 1 failed: failed to connect: ssh: connect to host localhost port 2222: Connection refused. Starting new attempt...
2025-04-18T11:04:32.316862374+01:00 [INFO] setting state to 'reconnecting'
2025-04-18T11:04:32.316933366+01:00 [INFO] Trying to reconnect to ssh server... Attempt 2
2025-04-18T11:04:32.322810286+01:00 [ERROR] Reconnect attempt 2 failed: failed to connect: ssh: connect to host localhost port 2222: Connection refused. Starting new attempt...
2025-04-18T11:04:32.322956977+01:00 [INFO] setting state to 'reconnecting'
2025-04-18T11:04:32.323024903+01:00 [INFO] Trying to reconnect to ssh server... Attempt 3
2025-04-18T11:04:32.329814073+01:00 [ERROR] Reconnect attempt 3 failed: failed to connect: ssh: connect to host localhost port 2222: Connection refused. Starting new attempt...
2025-04-18T11:04:32.329896185+01:00 [ERROR] Failed to reconnect to after 3 attempts, giving up
2025-04-18T11:04:32.3299581+01:00 [INFO] setting state to 'reconnect exhausted'
2025-04-18T11:04:32.338381809+01:00 [INFO] Creating a descriptor pool for at most 16 sets
2025-04-18T11:04:33.710895458+01:00 [INFO] Creating a descriptor pool for at most 16 sets
2025-04-18T11:04:33.71792876+01:00 [INFO] Creating a descriptor pool for at most 256 sets
2025-04-18T11:04:33.734534329+01:00 [INFO] Creating a descriptor pool for at most 256 sets
```

and after I click "Reconnect":
```
2025-04-18T11:04:51.094353906+01:00 [ERROR] client:unhandled ssh message name:UpdateProject
2025-04-18T11:04:51.19578317+01:00 [INFO] Using GPU: GpuSpecs { is_software_emulated: false, device_name: "AMD Radeon 780M (RADV GFX1103_R1)", driver_name: "radv", driver_info: "Mesa 24.2.8-1ubuntu1~24.04.1" }
2025-04-18T11:04:51.206615151+01:00 [INFO] shutting down ssh processes
2025-04-18T11:04:56.197306002+01:00 [ERROR] attempted to get project environment for a non-local worktree at "/home/atlas_universe/atlas_universe/apps/maze-portal"
2025-04-18T11:04:56.198955067+01:00 [INFO] perform;
2025-04-18T11:04:56.199025017+01:00 [INFO] perform;
2025-04-18T11:04:56.199116326+01:00 [INFO] authenticate;
2025-04-18T11:04:56.199213736+01:00 [INFO] perform;
2025-04-18T11:04:56.199317879+01:00 [INFO] authenticate;
2025-04-18T11:04:56.199402897+01:00 [INFO] perform;
2025-04-18T11:04:56.199489858+01:00 [INFO] authenticate;
2025-04-18T11:04:56.199583972+01:00 [INFO] read_command;
2025-04-18T11:04:56.199667557+01:00 [INFO] authenticate;
2025-04-18T11:04:56.199750771+01:00 [INFO] read_command;
2025-04-18T11:04:56.199832833+01:00 [INFO] perform;
2025-04-18T11:04:56.199943638+01:00 [INFO] authenticate;
2025-04-18T11:04:56.200026462+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.200100649+01:00 [INFO] read_command;
2025-04-18T11:04:56.200176009+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.200251218+01:00 [INFO] read_command;
2025-04-18T11:04:56.200346114+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.200446049+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.200546055+01:00 [INFO] read_command;
2025-04-18T11:04:56.200610284+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.200790067+01:00 [INFO] socket reader;
2025-04-18T11:04:56.200850879+01:00 [INFO] read_socket;
2025-04-18T11:04:56.200913435+01:00 [INFO] socket reader;
2025-04-18T11:04:56.200968547+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201024441+01:00 [INFO] socket reader;
2025-04-18T11:04:56.201096134+01:00 [INFO] read_socket;
2025-04-18T11:04:56.20114794+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201211017+01:00 [INFO] socket reader;
2025-04-18T11:04:56.201259497+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201316693+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201377054+01:00 [INFO] socket reader;
2025-04-18T11:04:56.201437847+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201489573+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201539525+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201605848+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201683592+01:00 [INFO] read_socket;
2025-04-18T11:04:56.201743553+01:00 [INFO] perform;
2025-04-18T11:04:56.201799687+01:00 [INFO] authenticate;
2025-04-18T11:04:56.201851694+01:00 [INFO] read_socket;
2025-04-18T11:04:56.20190904+01:00 [INFO] perform;
2025-04-18T11:04:56.201974622+01:00 [INFO] authenticate;
2025-04-18T11:04:56.202042247+01:00 [INFO] read_command;
2025-04-18T11:04:56.202124079+01:00 [INFO] perform;
2025-04-18T11:04:56.202176486+01:00 [INFO] authenticate;
2025-04-18T11:04:56.202232459+01:00 [INFO] read_command;
2025-04-18T11:04:56.202288974+01:00 [INFO] read_socket;
2025-04-18T11:04:56.20234089+01:00 [INFO] read_socket;
2025-04-18T11:04:56.202407353+01:00 [INFO] perform;
2025-04-18T11:04:56.202474087+01:00 [INFO] authenticate;
2025-04-18T11:04:56.202537434+01:00 [INFO] read_socket;
2025-04-18T11:04:56.202598518+01:00 [INFO] perform;
2025-04-18T11:04:56.202673366+01:00 [INFO] authenticate;
2025-04-18T11:04:56.202731795+01:00 [INFO] read_command;
2025-04-18T11:04:56.202783911+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.202844043+01:00 [INFO] read_command;
2025-04-18T11:04:56.202901629+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.202958455+01:00 [INFO] read_command;
2025-04-18T11:04:56.203030658+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.203086612+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.203155309+01:00 [INFO] send_secondary_commands;
2025-04-18T11:04:56.203243202+01:00 [INFO] socket reader;
2025-04-18T11:04:56.203303835+01:00 [INFO] read_socket;
2025-04-18T11:04:56.2033606+01:00 [INFO] socket reader;
2025-04-18T11:04:56.203424087+01:00 [INFO] read_socket;
2025-04-18T11:04:56.203475823+01:00 [INFO] socket reader;
2025-04-18T11:04:56.203539842+01:00 [INFO] read_socket;
2025-04-18T11:04:56.203604462+01:00 [INFO] read_socket;
2025-04-18T11:04:56.203635419+01:00 [INFO] read_socket;
2025-04-18T11:04:56.20367397+01:00 [INFO] read_socket;
2025-04-18T11:04:56.203712532+01:00 [INFO] read_socket;
2025-04-18T11:04:56.203745132+01:00 [INFO] read_socket;
2025-04-18T11:04:56.203881345+01:00 [INFO] socket reader;
2025-04-18T11:04:56.203914065+01:00 [INFO] read_socket;
2025-04-18T11:04:56.20394913+01:00 [INFO] read_socket;
2025-04-18T11:04:56.203980248+01:00 [INFO] socket reader;
2025-04-18T11:04:56.204011816+01:00 [INFO] read_socket;
2025-04-18T11:04:56.204041221+01:00 [INFO] read_socket;
2025-04-18T11:04:56.204079662+01:00 [INFO] read_socket;
2025-04-18T11:04:56.204190708+01:00 [INFO] read_socket;
2025-04-18T11:04:56.204306783+01:00 [INFO] read_socket;
2025-04-18T11:04:56.205961258+01:00 [ERROR] Failed to authenticate provider: Amazon Bedrock: DBus error service error org.freedesktop.zbus.Error: The name org.freedesktop.secrets was not provided by any .service files
2025-04-18T11:04:56.206011331+01:00 [ERROR] Failed to authenticate provider: DeepSeek: DBus error service error org.freedesktop.zbus.Error: The name org.freedesktop.secrets was not provided by any .service files
2025-04-18T11:04:56.206044132+01:00 [ERROR] Failed to authenticate provider: Google AI: DBus error service error org.freedesktop.zbus.Error: The name org.freedesktop.secrets was not provided by any .service files
2025-04-18T11:04:56.206079307+01:00 [ERROR] Failed to authenticate provider: Mistral: DBus error service error org.freedesktop.zbus.Error: The name org.freedesktop.secrets was not provided by any .service files
2025-04-18T11:04:56.206112368+01:00 [ERROR] Failed to authenticate provider: OpenAI: DBus error service error org.freedesktop.zbus.Error: The name org.freedesktop.secrets was not provided by any .service files
```

<!-- Before Submitting, did you:
  1. Include settings.json, keymap.json, .editorconfig if relevant?
  2. Check your Zed.log for relevant errors? (please include!)
  3. Click Preview to ensure everything looks right?
  4. Hide videos, large images and logs in ``` inside collapsible blocks:

<details><summary>click to expand</summary>

```json

```
</details>
-->


### Zed Version and System Specs

Zed: v0.182.11 (Zed) 
OS: Linux Wayland neon 24.04
Memory: 30.7 GiB
Architecture: x86_64
GPU: AMD Radeon 780M (RADV GFX1103_R1) || radv || Mesa 24.2.8-1ubuntu1~24.04.1

--- a/crates/extension_host/src/extension_host.rs
+++ b/crates/extension_host/src/extension_host.rs
@@ -1682,12 +1682,15 @@ impl ExtensionStore {
 
     pub fn register_ssh_client(&mut self, client: Entity<SshRemoteClient>, cx: &mut Context<Self>) {
         let connection_options = client.read(cx).connection_options();
-        if self.ssh_clients.contains_key(&connection_options.ssh_url()) {
-            return;
+        let ssh_url = connection_options.ssh_url();
+
+        if let Some(existing_client) = self.ssh_clients.get(&ssh_url) {
+            if existing_client.upgrade().is_some() {
+                return;
+            }
         }
 
-        self.ssh_clients
-            .insert(connection_options.ssh_url(), client.downgrade());
+        self.ssh_clients.insert(ssh_url, client.downgrade());
         self.ssh_registered_tx.unbounded_send(()).ok();
     }
 }

